pkg/tso/keyspace_group_manager.go

```diff
@@ -521,61 +528,75 @@ func (kgm *KeyspaceGroupManager) updateKeyspaceGroup(group *endpoint.KeyspaceGro
-			return
-		}
-
-		uniqueName := fmt.Sprintf("%s-%05d", kgm.electionNamePrefix, group.ID)
-		uniqueID := memberutil.GenerateUniqueID(uniqueName)
-		log.Info("joining primary election",
-			zap.Uint32("keyspace-group-id", group.ID),
-			zap.String("participant-name", uniqueName),
-			zap.Uint64("participant-id", uniqueID))
-
-		// TODO: handle the keyspace group & TSO split logic.
-		participant := member.NewParticipant(kgm.etcdClient)
-		participant.InitInfo(
-			uniqueName, uniqueID, path.Join(kgm.tsoSvcRootPath, fmt.Sprintf("%05d", group.ID)),
-			primaryElectionSuffix, "keyspace group primary election", kgm.cfg.GetAdvertiseListenAddr())
-
-		// Only the default keyspace group uses the legacy service root path for LoadTimestamp/SyncTimestamp.
-		var (
-			tsRootPath string
-			storage    *endpoint.StorageEndpoint
-		)
-		if group.ID == mcsutils.DefaultKeyspaceGroupID {
-			tsRootPath = kgm.legacySvcRootPath
-			storage = kgm.legacySvcStorage
-		} else {
-			tsRootPath = kgm.tsoSvcRootPath
-			storage = kgm.tsoSvcStorage
-		}
-
-		am := NewAllocatorManager(kgm.ctx, group.ID, participant, tsRootPath, storage, kgm.cfg, true)
-
-		kgm.Lock()
-		group.KeyspaceLookupTable = make(map[uint32]struct{})
-		for _, kid := range group.Keyspaces {
-			group.KeyspaceLookupTable[kid] = struct{}{}
-			kgm.keyspaceLookupTable[kid] = group.ID
-		}
-		kgm.kgs[group.ID] = group
-		kgm.ams[group.ID] = am
-		kgm.Unlock()
-	} else {
+	// Not assigned to me. If this host/pod owns this keyspace group, it should resign.
+	if !kgm.isAssignedToMe(group) {
 		if group.ID == mcsutils.DefaultKeyspaceGroupID {
 			log.Info("resign default keyspace group membership",
 				zap.Any("default-keyspace-group", group))
 		}
-		// Not assigned to me. If this host/pod owns this keyspace group, it should resign.
 		kgm.deleteKeyspaceGroup(group.ID)
+		return
+	}
+	// If the keyspace group is already initialized, just update the meta.
+	if oldAM, oldGroup := kgm.state.getKeyspaceGroupMeta(group.ID); oldAM != nil {
+		log.Info("keyspace group already initialized, so update meta only",
+			zap.Uint32("keyspace-group-id", group.ID))
+		kgm.updateKeyspaceGroupMembership(oldGroup, group)
+		return
 	}
+	// If the keyspace group is not initialized, initialize it.
+	uniqueName := fmt.Sprintf("%s-%05d", kgm.electionNamePrefix, group.ID)
+	uniqueID := memberutil.GenerateUniqueID(uniqueName)
+	log.Info("joining primary election",
+		zap.Uint32("keyspace-group-id", group.ID),
+		zap.String("participant-name", uniqueName),
+		zap.Uint64("participant-id", uniqueID))
+	// Initialize the participant info to join the primary election.
+	participant := member.NewParticipant(kgm.etcdClient)
+	participant.InitInfo(
+		uniqueName, uniqueID, path.Join(kgm.tsoSvcRootPath, fmt.Sprintf("%05d", group.ID)),
+		primaryElectionSuffix, "keyspace group primary election", kgm.cfg.GetAdvertiseListenAddr())
+	// If the keyspace group is in split, we should ensure that the primary elected by the new keyspace group
```

Comment:
How do we force the constraint that the primaries of the split-from kg and split-to kg are on the same tso node? Do we reuse the logic of "nextLeader" in local tso allocators? Will this election algorithm change be in this pr or in separate pr?

Reply:
This is guaranteed by two constraints:

- Once a keyspace group is in the split state, its meta info could not be modified after #6316.
- `participant.SetPreCampaignChecker` below makes sure the primary could only be elected by the participant which owns the split source keyspace group primary also.